Issue#1100 Provide argument in AutoSklearnClassifier and AutoSklearnRegressor to pass callback function

bitsbuffer opened the pull request:
This PR has implemented the functionality as described in https://github.com/automl/auto-sklearn/issues/1100
which provides an argument "get_trials_callback" that accepts a callback object which will be executed after finishing a trial.

Test case to test this functionality is added at test/test_util/test_trials_callback.py

bitsbuffer commented:
Checkout the branch bitsbuffer:issue_1100_trials_callback from master instead development.

mfeurer commented on test/test_util/test_trials_callback.py:
```diff
@@ -0,0 +1,92 @@
+class VerifyTrialsCallBack(unittest.TestCase):
+
+    def test_trials_callback_execution(self):
+        base_dir = "./tmp"
```
Could you please use python's built-in `tempdir` module?

bitsbuffer replied:
Hi @mfeurer, it makes sense to allow for newly added callbacks in SMAC. Modified the code to support it.

mfeurer commented on test/test_util/test_trials_callback.py:
```diff
@@ -0,0 +1,92 @@
+                                   initial_configurations_via_metalearning=0,
+                                   per_run_time_limit=10,
+                                   memory_limit=1024,
+                                   tmp_folder=tmp_folder,
```
You don't need to provide a `tmp_folder` and an `output_folder` for the unit tests.

bitsbuffer replied:
modified the code to use tempfile module from Python.

mfeurer commented on test/test_util/test_trials_callback.py:
```diff
@@ -0,0 +1,92 @@
+        cls.fit(X_train, Y_train, X_test,Y_test)
+        trials = pd.read_csv(os.path.join(base_dir, "trials.csv"))
+        assert trials.shape[0] > 1
+        assert len(os.listdir(output_folder)) > 0
```
What's in the `output_folder`?

bitsbuffer replied:
I was actually checking if AutoSklearnClassifier produces any output as for each successful run it creates a run directory provided ensemble_size = 0. Removed that statement as it was not required in test case. The main purpose of the callback was to gather summary of trials irrespective of trial's run status.

mfeurer commented on test/test_util/test_trials_callback.py:
```diff
@@ -0,0 +1,92 @@
+                                   tmp_folder=tmp_folder,
+                                   delete_tmp_folder_after_terminate=False,
+                                   output_folder=output_folder,
+                                   delete_output_folder_after_terminate=False,
```
Could you please explain why you would like to keep the tmp and the output folder?

bitsbuffer replied:
Modified the code as it is not required.

mfeurer commented on test/test_util/test_trials_callback.py:
```diff
@@ -0,0 +1,92 @@
+        assert len(os.listdir(output_folder)) > 0
+        shutil.rmtree(base_dir)
+
+    def test_no_trials_callback(self):
```
Could you please apply the same comments as the ones I made to the above unit test?

bitsbuffer replied:
Modified the code as it is not required.

mfeurer commented on doc/index.rst:
```diff
@@ -96,7 +96,7 @@ reference to the following paper:
 
 
- Auto-Sklearn 2.0: The Next Generation, Feurer *et al.*, to appear (2020).
+ `Auto-Sklearn 2.0: The Next Generation <https://arxiv.org/abs/2007.04074>`_, Feurer *et al.*, (arXiv, 2020).
```
Thanks for fixing this!

mfeurer commented on autosklearn/estimators.py:
```diff
@@ -52,6 +54,7 @@ def __init__(
         metric=None,
         scoring_functions: Optional[List[Scorer]] = None,
         load_models: bool = True,
+        get_trials_callback: Optional[IncorporateRunResultCallback] = None
```
Could you please add the callback to the docstring so that it's rendered in the docs?

bitsbuffer replied:
Modified the code to add docstring.

mfeurer reviewed (commented):
Thanks a lot for the contribution and please excuse the long delay. I left some notes in the comments, could you please update your code on them?

I was also wondering whether it would make sense to already allow for other callbacks, too, in case we add them to SMAC, what do you think?

mfeurer commented:
Hey @bitsbuffer, please excuse the horrible delay. I have just approved the PR and made two small changes. However, there have been updates since the PR which now break your changes. Could you please rebase and update your PR, merging then will be happen quickly.

bitsbuffer commented:
@mfeurer Sure, I will take care of it.

bitsbuffer commented:
@mfeurer I've modified pull request which include rebase to latest development branch, also modified the test case to remove delete_output_folder_after_terminate=False.

mfeurer commented:
Thanks a lot @bitsbuffer, unfortunately, the new test isn't passing and there is an issue with the automatic style checker. Could you please have a look into these?

bitsbuffer commented:
@mfeurer Apologies for this. I have modified test_check_askl2_same_arguments_as_askl and code style as indicated by automatic style checker.

mfeurer commented:
Hey @bitsbuffer thanks a lot for the update. Unfortunately, there are still two complaints about unnecessary imports. Could you please have a look at them?

bitsbuffer commented:
Hi @mfeurer, probably I need to be more careful with this from the next time. I removed unused imports from estimator.py

mfeurer commented:
Thanks a lot for your contribution!

(and yes, we need to make the process of contributing much smoother)
